uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php

```diff
@@ -135,4 +145,300 @@ public function testDeleteDeletesTheLeaveRequestItsBalanceChangesAndDates() {
+    $contactID = 2;
+    $this->getleaveRequestService()->create([
+      'type_id' => 1,
+      'contact_id' => $contactID,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You can't create a Leave Request with this status
+   */
+  public function testCreateThrowsAnExceptionWhenTransitionStatusIsNotValidForNewLeaveRequest() {
+    //A leave contact trying to set a new leave request to approved status
+    $this->getleaveRequestService()->create([
+      'type_id' => 1,
+      'contact_id' => $this->leaveContact,
+      'status_id' => $this->leaveRequestStatuses['Approved']['id'],
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You can't create a Leave Request with this status
+   */
+  public function testCreateThrowsAnExceptionForAdminWhenTransitionStatusIsNotValidForNewLeaveRequest() {
+    //An admin trying to set a new leave request to rejected status
+    $this->getLeaveRequestServiceWhenCurrentUserIsAdmin()->create([
+      'type_id' => 1,
+      'contact_id' => 5,
+      'status_id' => $this->leaveRequestStatuses['Rejected']['id'],
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You can't create a Leave Request with this status
+   */
+  public function testCreateThrowsAnExceptionForLeaveApproverWhenTransitionStatusIsNotValidForNewLeaveRequest() {
+    //A leave manager trying to set a new leave request to cancelled status
+    $this->getLeaveRequestServiceWhenCurrentUserIsLeaveManager()->create([
+      'type_id' => 1,
+      'contact_id' => 5,
+      'status_id' => $this->leaveRequestStatuses['Cancelled']['id'],
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
```

Comment:
All the new tests here can probably be simplified a lot if you mock the services inject into the LeaveRequest Service. For example, instead of passing all these params and making sure that a user is admin and bla bla bla, just you can mock the LeaveRequestStatusMatrix to simply return false.

This also means you actually only need one test for each error instead of 3 (one for current use, one for manager and one for admin). Something like `testCreateThrowsAnExceptionIfTheStatusTransitionIsNotValid()`